src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -1953,117 +1965,85 @@ public void updateLatestRealtimeTask(
      *
      * @param detectorId detector id
-     * @param newState new state
+     * @param state new state
      * @param rcfTotalUpdates rcf total updates
      * @param detectorIntervalInMinutes detector interval in minutes
      * @param error error
      * @param listener action listener
      */
-    public void updateLatestRealtimeTask(
+    public void updateLatestRealtimeTaskOnCoordinatingNode(
         String detectorId,
-        String newState,
+        String state,
         Long rcfTotalUpdates,
         Long detectorIntervalInMinutes,
         String error,
         ActionListener<UpdateResponse> listener
     ) {
         Float initProgress = null;
-        String state = null;
+        String newState = null;
         // calculate init progress and task state with RCF total updates
         if (detectorIntervalInMinutes != null && rcfTotalUpdates != null) {
-            state = ADTaskState.INIT.name();
+            newState = ADTaskState.INIT.name();
             if (rcfTotalUpdates < NUM_MIN_SAMPLES) {
                 initProgress = (float) rcfTotalUpdates / NUM_MIN_SAMPLES;
             } else {
-                state = ADTaskState.RUNNING.name();
+                newState = ADTaskState.RUNNING.name();
                 initProgress = 1.0f;
             }
         }
         // Check if new state is not null and override state calculated with rcf total updates
-        if (newState != null) {
-            state = newState;
+        if (state != null) {
+            newState = state;
         }
 
         error = Optional.ofNullable(error).orElse("");
-        if (!adTaskCacheManager.isRealtimeTaskChanged(detectorId, state, initProgress, error)) {
+        if (!adTaskCacheManager.isRealtimeTaskChanged(detectorId, newState, initProgress, error)) {
+            // If task not changed, no need to update, just return
+            listener.onResponse(null);
             return;
         }
-        if (ADTaskState.STOPPED.name().equals(state)) {
-            adTaskCacheManager.removeRealtimeTaskCache(detectorId);
-        }
         Map<String, Object> updatedFields = new HashMap<>();
         updatedFields.put(COORDINATING_NODE_FIELD, clusterService.localNode().getId());
         if (initProgress != null) {
             updatedFields.put(INIT_PROGRESS_FIELD, initProgress);
             updatedFields.put(ESTIMATED_MINUTES_LEFT_FIELD, Math.max(0, NUM_MIN_SAMPLES - rcfTotalUpdates) * detectorIntervalInMinutes);
         }
-        if (state != null) {
-            updatedFields.put(STATE_FIELD, state);
+        if (newState != null) {
+            updatedFields.put(STATE_FIELD, newState);
         }
         if (error != null) {
             updatedFields.put(ERROR_FIELD, error);
         }
-        Float newInitProgress = initProgress;
-        String newError = error;
+        Float finalInitProgress = initProgress;
+        String finalError = error;
+        String finalNewState = newState;
```

Comment:
why do you need to do these assignments?

Reply:
Variable used in lambda expression should be final or effectively final

Reply:
Can you add a comment for this?

Reply:
Sure